### ISSUE TITLE
Reading serialized wavefunctions for CASSCF

### ISSUE THREAD

**Issue by bsenjean** [psi4/psi4#1871]
On issue #1861 , reading serialized wavefunctions for CASCI/detci computations was solved.

However, I wonder if it was also solved for CASSCF calculations, as I get the following error when restarting a CASSCF calculation with a given input wavefunction.

Here is an example:

```
import psi4
molecule_string = """
O 1.56850 0.10589 0.00001
N -1.39559 -0.02156 0.00004
H -1.62981 0.96110 -0.10622
H -1.86277 -0.51254 -0.75597
H -1.83355 -0.33077 0.86231
H 0.60674 -0.03396 -0.00063
H 1.94052 -0.78000 0.00022
"""
add_input = "\nsymmetry c1\nno_reorient\nno_com"
mol = psi4.geometry(molecule_string + add_input)
N_active  = 11 # Define the total number of active MOs.
N_frozen  = 4 # Define the total number of MOs in the active space
N_virtual = 30 # Define the total number of virtual MOs that will be discarded
psi4.set_options({ 'ACTIVE' :      [ N_active ],\
                   'FROZEN_DOCC' : [ N_frozen ],\
                   'FROZEN_UOCC' : [ N_virtual],\
                   'DIAG_METHOD' : 'DAVIDSON',\
                   'CI_MAXITER'  : 200\
                 })

# assuming you have stored you scf_wfn into water-ammonia.npy:
original_wfn = psi4.core.Wavefunction.from_file("water-ammonia")

print("HF energy:",original_wfn.energy())
CASCI_energy = psi4.energy('fci', molecule=mol, ref_wfn=original_wfn,  return_wfn=False)
print("CASCI energy:",CASCI_energy)
CASSCF_energy = psi4.energy('casscf', molecule=mol, ref_wfn=original_wfn,  return_wfn=False)
print("CASSCF :",CASSCF_energy)
```

This outputs:
```HF energy : -130.74589276437956
CAS energy : -130.7730891288199
```

and then the error:
```
   ==> Setting up CI strings <==

    There are 462 alpha and 462 beta strings
    The CI space requires 213444 (2.13E+05) determinants and 1 blocks

   ==> Setting up MCSCF integrals <==

Could not find requested basisset (DF_BASIS_SCF).Traceback (most recent call last):
  File "PSI4_CAS_sim.py", line 89, in <module>
    CASSCF_energy = psi4.energy('casscf', molecule=mol, ref_wfn=original_wfn,  return_wfn=False)
  File "/Users/bsenjean/Documents/Git_repositories/psi4/objdir/stage/lib/psi4/driver/driver.py", line 570, in energy
    wfn = procedures['energy'][lowername](lowername, molecule=molecule, **kwargs)
  File "/Users/bsenjean/Documents/Git_repositories/psi4/objdir/stage/lib/psi4/driver/procrouting/proc.py", line 4804, in run_detcas
    ciwfn = mcscf.mcscf_solver(ref_wfn)
  File "/Users/bsenjean/Documents/Git_repositories/psi4/objdir/stage/lib/psi4/driver/procrouting/mcscf/mcscf_solver.py", line 100, in mcscf_solver
    mcscf_obj = ciwfn.mcscf_object()
RuntimeError: 
Fatal Error: Wavefunction::get_basisset: Requested basis set (DF_BASIS_SCF) was not set!

Error occurred in file: /Users/bsenjean/Documents/Git_repositories/psi4/psi4/src/psi4/libmints/wavefunction.cc on line: 670
The most recent 1 function calls were:

```

**Comment by hokru** [comment]
Looks like a bug.

It gets stuck here:
https://github.com/psi4/psi4/blob/master/psi4/src/psi4/detci/ints.cc#L369

**Comment by hokru** [comment]
Not sure how this would be best fixed, but it can be manually mitigated by setting the basis set yourself.
You are using conventional integrals and no density fitting basis is needed, but the constructor in the code wants an object. This sets an empty basis set: 
`original_wfn.set_basisset("DF_BASIS_SCF", core.BasisSet.zero_ao_basis_set())`

Let us know if works for you.

**Comment by JonathonMisiewicz** [comment]
I can't even run this input file. What basis set did you use?

**Comment by hokru** [comment]
You would need to make your own wfn file first.

The serial-wfn test seems fine for testing:

```
molecule mol {
0 1
O 0.0 0.0 0.0
H 1.0 0.0 0.0
H -.7 0.7 0.0
symmetry c1
}
memory 4 GiB
set {
    basis sto-3g
    guess sad
 active [7]
}
# make a wavefunction
e, wfn_old = energy('hf', return_wfn=True)

# write the wavefunction to file
wfn_old.to_file('my_wfn')

# read wavefunction from file
wfn_new = Wavefunction.from_file('my_wfn')
#wfn_new.set_basisset("DF_BASIS_SCF", core.BasisSet.zero_ao_basis_set())
energy('casscf',ref_wfn=wfn_new)
```

I notice now that I get a segfault at the `Properties computed using the CASSCF density matrix`

**Comment by JonathonMisiewicz** [comment]
Let me rephrase: I tried to make my own wavefunction file, but when I succeed in that, I get a different error when setting up the active space, likely because the number of MOs in the basis set I'm picking isn't consistent with the active space defined in the input file.

Anyways, this bug is the result of the exact same deep design issue in the Wavefunction class that led to #1851. **I don't know what the Wavefunction "contract" is, and wavefunction serialization disagrees about what the contract is with every single module that's having these problems.**

#1851 : Some modules assume that every wavefunction must have a PSIO object. Wavefunction serialization disagrees.
bsenjean's original report : The JK object constructors force callers to pass in an auxiliary basis set, even when the JK object will never use it. `detci` responds by assuming that every reference wavefunction must have a `DF_BASIS_SCF` set. Wavefunction serialization disagrees. Holger's previous analysis is right on for the mechanism of this.
Holger's new report : The properties code [assumes that every wavefunction has a matrix factory](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libmints/oeprop.cc#L88) and uses that to create an [OperatorSymmetry object](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libmints/oeprop.cc#L1256). Wavefunction serialization disagrees. It doesn't attach a matrix factory, so libmints is left trying to create matrices out of `nullptr`. That segmentation faults, unsurprisingly.

I don't have time to deal with these, because I have the gradient code to worry about, and I'm sure nobody else has time because _getting 1.4 out soon is kind of important_. I will say that this is not a good state of affairs.

**Comment by bsenjean** [comment]
Sorry I realize my example was incomplete.
I used the following options:

```
    psi4.set_options({'basis': 'STO-3G-DECON',
                      'reference': 'rohf',
                      'scf_type': 'direct',
                      'guess': 'sad',
                      'puream':true})
```

I then used @hokru 's trick to set the ```DF_BASIS_SCF``` object. I modified the active space (using ```RESTRICTED_DOCC``` instead of ```FROZEN_DOCC``` and only 4 active orbitals) and I also get the Segmentation Fault in the property section.

But the energy is there, which is already good enough for me right now, so thanks for the tip !

**Comment by JonathonMisiewicz** [comment]
That's good to hear. Thanks for the report, but I think it will be some time before further progress happens. This is a deep issue, at least from my understanding.
